src/test_memcachedsolution.cpp

```diff
@@ -0,0 +1,864 @@
+  {
+    return _store->get_data(_table,
+                            key,
+                            data,
+                            cas,
+                            DUMMY_TRAIL_ID);
+  }
+
+  /// Helper method for deleting data from memcached for the test's default key.
+  Store::Status delete_data()
+  {
+    return _store->delete_data(_table, _key, DUMMY_TRAIL_ID);
+  }
+
+  TopologyNeutralMemcachedStore* _store;
+
+  /// Use shared pointers for managing the instances so that the memory gets
+  /// freed when the vector is cleared.
+  static std::vector<std::shared_ptr<MemcachedInstance>> _memcached_instances;
+  static std::vector<std::shared_ptr<AstaireInstance>> _astaire_instances;
+
+  /// Tests that use this fixture use a monotonically incrementing numerical key
+  /// (so that tests are isolated from each other). This variable stores the
+  /// next key to use.
+  static unsigned int _next_key;
+
+  // The table and key to use for the test. The table is the same in all tests.
+  const static std::string _table;
+  std::string _key;
+};
+
+std::vector<std::shared_ptr<MemcachedInstance>> BaseMemcachedSolutionTest::_memcached_instances;
+std::vector<std::shared_ptr<AstaireInstance>> BaseMemcachedSolutionTest::_astaire_instances;
+
+unsigned int BaseMemcachedSolutionTest::_next_key;
+const std::string BaseMemcachedSolutionTest::_table = "test_table";
+
+/// A class for killing and restoring a memcached instance. Used as a template
+/// along with MemcachedRestarter in the tests below.
+class MemcachedKiller
+{
+  static void fail_memcached_instance(std::shared_ptr<MemcachedInstance> instance)
+  {
+    instance->kill_instance();
+  }
+
+  static void restore_memcached_instance(std::shared_ptr<MemcachedInstance> instance)
+  {
+    instance->start_instance();
+    EXPECT_TRUE(instance->wait_for_instance());
+  }
+};
+
+/// A class for restarting a memcached instance. Used as a template along with
+/// MemcachedKiller in the tests below.
+class MemcachedRestarter
+{
+  static void fail_memcached_instance(std::shared_ptr<MemcachedInstance> instance)
+  {
+    instance->restart_instance();
+    EXPECT_TRUE(instance->wait_for_instance());
+  }
+
+  static void restore_memcached_instance(std::shared_ptr<MemcachedInstance> instance)
+  {
+  }
+};
+
+typedef ::testing::Types<MemcachedKiller, MemcachedRestarter> MemcachedFailHelpers;
+
+/// A fixture for testing a deployment with 1 Astaire and 2 memcacheds.
+/// Templated so that we can test that behaviour is the same in the scenarios
+/// where a memcached dies and where a memcached bounces.
+template <class T>
+class SimpleMemcachedSolutionTest : public BaseMemcachedSolutionTest
+{
+  static void SetUpTestCase()
+  {
+    _next_key = std::rand();
+    create_and_start_memcached_instances(2);
+    create_and_start_astaire_instances(1);
+    EXPECT_TRUE(wait_for_instances());
+  }
+};
+
+TYPED_TEST_CASE(SimpleMemcachedSolutionTest, MemcachedFailHelpers);
+
```

Comment:
Could you add a big "Testcases start here" banner to make it easier to find your place in the file when quickly scanning it?